### PR TITLE
Add more detailed timing metrics to orchestrator

### PR DESCRIPTION
**Pull request by sjmudd**
This rather large patch brings in code from booking.com which does the following:

Signal handling:
* handle SIGTERM

discoverInstance():
* Add timing metrics: talking to instance, backend timings and total
  time for each discovery using vendor/github.com/sjmudd/stopwatch
  for managing timing counters.

Add new api calls:
* /api/discovery-metrics-raw/:seconds
* /api/discovery-metrics-aggregated/:seconds
* /api/backend-query-metrics-raw/:seconds
* /api/backend-query-metrics-aggregated/:seconds

New configuration setting:
* DiscoveryCollectionRetentionSeconds which determines the amount of time
  individual discovery statistics are kept for.

checkMaxScale:
* move maxscale checks out of inst.ReadTopologyInstanceBufferable()

Basically the change adds more detailed metrics which can be seen in logging but also makes these metrics retrievable by API calls to orchestrator. Both raw and aggregated forms are provided to give the user to choose what works best.

These changes were added to our code to help us identify performance problems on our large setup and have been running for some time.

- [x] contributed code is using same conventions as original code
- [x] code is formatted via `gofmt` (please avoid `goimports`)
- [x] code is built via `./build.sh`
- [x] code is tested via `go test ./go/...`

### CODE REVIEW

File: go/logic/orchestrator.go [hunk 1]
```diff
@@ -71,13 +76,21 @@ func acceptSignals() {
...
-				config.Reload()
 				inst.AuditOperation("reload-configuration", nil, "Triggered via SIGHUP")
+				config.Reload()
+				discoveryMetrics.SetExpirePeriod(time.Duration(config.Config.DiscoveryCollectionRetentionSeconds) * time.Second)
+			case syscall.SIGTERM:
+				log.Debugf("Received SIGTERM. Shutting down orchestrator")
+				discoveryMetrics.StopAutoExpiration()
+				// probably should poke other go routines to stop cleanly here ...
+				inst.AuditOperation("shutdown", nil, "Triggered via SIGTERM")
+				os.Exit(0)
```
Comment (shlomi-noach) [inline go/logic/orchestrator.go:93]: :+1:

File: go/logic/orchestrator.go [hunk 2]
```diff
@@ -71,13 +76,21 @@ func acceptSignals() {
 	c := make(chan os.Signal, 1)
 
 	signal.Notify(c, syscall.SIGHUP)
+	signal.Notify(c, syscall.SIGTERM)
```
Comment (shlomi-noach) [inline go/logic/orchestrator.go:79]: :+1:

File: go/config/config.go [hunk 3]
```diff
@@ -84,6 +84,7 @@ type Configuration struct {
 	DiscoveryMaxConcurrency                      uint     // Number of goroutines doing hosts discovery
 	DiscoveryQueueCapacity                       uint     // Buffer size of the discovery queue. Should be greater than the number of DB instances being discovered
 	DiscoveryQueueMaxStatisticsSize              int      // The maximum number of individual secondly statistics taken of the discovery queue
+	DiscoveryCollectionRetentionSeconds          uint     // Number of seconds to retain the discovery collection information
```
Comment (shlomi-noach) [inline go/config/config.go:87]: :+1:

File: go/config/config.go [hunk 4]
```diff
@@ -251,6 +252,7 @@ func newConfiguration() *Configuration {
 		DiscoveryMaxConcurrency:                      300,
 		DiscoveryQueueCapacity:                       100000,
 		DiscoveryQueueMaxStatisticsSize:              120,
+		DiscoveryCollectionRetentionSeconds:          120,
```
Comment (shlomi-noach) [inline go/config/config.go:255]: Does this mean `DiscoveryCollectionRetentionSeconds` is ON by default? What is the change in behavior I can expect on my own setup as I deploy this?

File: docs/using-the-web-api.md [hunk 5]
```diff
@@ -62,9 +62,13 @@ The following bulk retrieval rules are intended for allowing the information ins
...
 * `/api/bulk-promotion-rules`: provide a json list of instance promotion rules in the form of Hostname Port PromotionRule
 
-The following monitoring URLs provide more timing information for discovery processes
+The following URLs for collecting monitoring information
+* `/api/discovery-metrics-raw/:seconds`: list the raw discovery metrics for the last specified number of seconds. Raw discovery metric data is kept for DiscoveryCollectionRetentionSeconds seconds (default: 120)
+* `/api/discovery-metrics-aggregated/:seconds`: provide aggregated data for the discovery metrics over the last specified number of seconds.
 * `/api/discovery-queue-metrics-raw/:seconds`: return a list of active and queued number discovery keys over the requested number of seconds
 * `/api/discovery-queue-metrics-aggregated/:seconds`: return aggregated data based on the raw data mentioned above
+* `/api/backend-query-metrics-raw/:seconds`: return the raw metrics on queries sent to the backend in the period specified
+* `/api/backend-query-metrics-aggregated/:seconds`: return a set of aggregated values based on the raw numbers above
```
Comment (shlomi-noach) [inline docs/using-the-web-api.md:71]: :+1:

File: go/app/http.go [hunk 6]
```diff
@@ -35,19 +37,22 @@ import (
...
+func Http(continuousDiscovery bool) {
 	promptForSSLPasswords()
 	process.ContinuousRegistration(process.OrchestratorExecutionHttpMode, "")
 
 	martini.Env = martini.Prod
 	if config.Config.ServeAgentsHttp {
 		go agentsHttp()
 	}
-	standardHttp(discovery)
+	standardHttp(continuousDiscovery)
```
Comment (shlomi-noach) [inline go/app/http.go:55]: :+1:

File: go/app/http.go [hunk 7]
```diff
@@ -35,19 +37,22 @@ import (
...
 
+const discoveryMetricsName = "DISCOVERY_METRICS"
+
 var sslPEMPassword []byte
 var agentSSLPEMPassword []byte
+var discoveryMetrics *collection.Collection
 
 // Http starts serving
-func Http(discovery bool) {
+func Http(continuousDiscovery bool) {
```
Comment (shlomi-noach) [inline go/app/http.go:47]: :+1:

File: go/app/http.go [hunk 8]
```diff
@@ -114,7 +119,11 @@ func standardHttp(discovery bool) {
 
 	inst.SetMaintenanceOwner(process.ThisHostname)
 
-	if discovery {
+	if continuousDiscovery {
+		// start to expire metric collection info
+		discoveryMetrics = collection.CreateOrReturnCollection(discoveryMetricsName)
+		discoveryMetrics.SetExpirePeriod(time.Duration(config.Config.DiscoveryCollectionRetentionSeconds) * time.Second)
+
```
Comment (shlomi-noach) [inline go/app/http.go:126]: :+1:

File: go/http/api.go [hunk 9]
```diff
@@ -1607,7 +1646,7 @@ func (this *HttpAPI) DiscoveryQueueMetricsRaw(params martini.Params, r render.Re
 
 	queue := discovery.CreateOrReturnQueue("DEFAULT")
 	metrics := queue.DiscoveryQueueMetrics(seconds)
-	log.Debugf("Raw data: %+v", metrics)
+	log.Debugf("DiscoveryQueueMetricsRaw data: %+v", metrics)
```
Comment (shlomi-noach) [inline go/http/api.go:1649]: :+1:

File: go/http/api.go [hunk 10]
```diff
@@ -1619,13 +1658,49 @@ func (this *HttpAPI) DiscoveryQueueMetricsAggregated(params martini.Params, r re
 	seconds, err := strconv.Atoi(params["seconds"])
 	log.Debugf("DiscoveryQueueMetricsAggregated: seconds: %d", seconds)
 	if err != nil {
-		r.JSON(200, &APIResponse{Code: ERROR, Message: "Unable to generate discovery queue  aggregated metrics"})
+		r.JSON(200, &APIResponse{Code: ERROR, Message: "Unable to generate discovery queue aggregated metrics"})
```
Comment (shlomi-noach) [inline go/http/api.go:1661]: :+1:

File: go/http/api.go [hunk 11]
```diff
@@ -2525,8 +2600,12 @@ func (this *HttpAPI) RegisterRequests(m *martini.ClassicMartini) {
 	this.registerRequest(m, "bulk-promotion-rules", this.BulkPromotionRules)
 
 	// Monitoring
+	this.registerRequest(m, "discovery-metrics-raw/:seconds", this.DiscoveryMetricsRaw)
+	this.registerRequest(m, "discovery-metrics-aggregated/:seconds", this.DiscoveryMetricsAggregated)
```
Comment (shlomi-noach) [inline go/http/api.go:2604]: :+1:

File: go/http/api.go [hunk 12]
```diff
@@ -2525,8 +2600,12 @@ func (this *HttpAPI) RegisterRequests(m *martini.ClassicMartini) {
...
 	this.registerRequest(m, "discovery-queue-metrics-raw/:seconds", this.DiscoveryQueueMetricsRaw)
 	this.registerRequest(m, "discovery-queue-metrics-aggregated/:seconds", this.DiscoveryQueueMetricsAggregated)
+	this.registerRequest(m, "backend-query-metrics-raw/:seconds", this.BackendQueryMetricsRaw)
+	this.registerRequest(m, "backend-query-metrics-aggregated/:seconds", this.BackendQueryMetricsAggregated)
```
Comment (shlomi-noach) [inline go/http/api.go:2608]: :+1:

File: go/logic/orchestrator.go [hunk 13]
```diff
@@ -142,16 +166,43 @@ func discoverInstance(instanceKey inst.InstanceKey) {
 	discoveriesCounter.Inc(1)
 
 	// First we've ever heard of this instance. Continue investigation:
-	instance, err = inst.ReadTopologyInstanceBufferable(&instanceKey, config.Config.BufferInstanceWrites)
+	instance, err = inst.ReadTopologyInstanceBufferable(&instanceKey, config.Config.BufferInstanceWrites, latency)
```
Comment (shlomi-noach) [inline go/logic/orchestrator.go:169]: :+1:

File: go/logic/orchestrator.go [hunk 14]
```diff
@@ -142,16 +166,43 @@ func discoverInstance(instanceKey inst.InstanceKey) {
...
+			BackendLatency:  latency.Elapsed("backend"),
+			InstanceLatency: latency.Elapsed("instance"),
+			Err:             err,
+		})
+		log.Warningf("discoverInstance(%+v) instance is nil in %.3fs (Backend: %.3fs, Instance: %.3fs), error=%+v",
+			instanceKey,
+			latency.ElapsedSeconds("total"),
+			latency.ElapsedSeconds("backend"),
+			latency.ElapsedSeconds("instance"),
+			err)
```
Comment (shlomi-noach) [inline go/logic/orchestrator.go]: Is it possible that between the `discoveryMetrics.Append(...)` call and this line, the returned value of `ElapsedSeconds(...)` are incompatible? Would this be confusing?

File: go/logic/orchestrator.go [hunk 15]
```diff
@@ -142,16 +166,43 @@ func discoverInstance(instanceKey inst.InstanceKey) {
...
+		InstanceLatency: latency.Elapsed("instance"),
+		Err:             nil,
+	})
+	log.Debugf("Discovered host: %+v, master: %+v, version: %+v in %.3fs (Backend: %.3fs, Instance: %.3fs)",
+		instance.Key,
+		instance.MasterKey,
+		instance.Version,
+		latency.ElapsedSeconds("total"),
+		latency.ElapsedSeconds("backend"),
+		latency.ElapsedSeconds("instance"))
```
Comment (shlomi-noach) [inline go/logic/orchestrator.go]: 👍

File: vendor/github.com/sjmudd/stopwatch/named_stopwatch.go [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+// ElapsedMilliSeconds returns the elapsed time in milliseconds of
+// the named stopwatch if it exists or 0.
+func (ns *NamedStopwatch) ElapsedMilliSeconds(name string) float64 {
+	if s, ok := ns.stopwatches[name]; ok {
+		return s.ElapsedMilliSeconds()
+	}
+	return float64(0)
+}
```
Comment (shlomi-noach) [inline vendor/github.com/sjmudd/stopwatch/named_stopwatch.go:175]: I see no synchronization code here. What happens on concurrent writes? Is it possible to call `Keys()` or `Reset()` while calling `AddMany()` or `Delete()`?

File: go/collection/collection.go [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+	c.Lock()
+	defer c.Unlock()
+	// we don't want to add nil metrics
+	if c == nil {
+		return errors.New("Collection.Append: c == nil")
+	}
+	c.collection = append(c.collection, m)
+
+	return nil
+}
```
Comment (shlomi-noach) [inline go/collection/collection.go:292]: :+1:

File: go/collection/collection_test.go [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+		t.Errorf("TestAppend: len(Metrics) = %d, expecting %d", len(c.Metrics()), 0)
+	}
+	for _, v := range []int{1, 2, 3} {
+		tm := &testMetric{}
+		c.Append(tm)
+		if len(c.Metrics()) != v {
+			t.Errorf("TestExpirePeriod: len(Metrics) = %d, expecting %d", len(c.Metrics()), v)
+		}
+	}
+}
```
Comment (shlomi-noach) [inline go/collection/collection_test.go:104]: :+1:

File: go/discovery/aggregated.go [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+// AggregatedSince returns a large number of aggregated metrics
+// based on the raw metrics collected since the given time.
+func AggregatedSince(c *collection.Collection, t time.Time) (AggregatedDiscoveryMetrics, error) {
+	results, err := c.Since(t)
+	if err != nil {
+		return AggregatedDiscoveryMetrics{}, err
+	}
+
+	return aggregate(results), nil
+}
```
Comment (shlomi-noach) [inline go/discovery/aggregated.go:175]: :+1:

File: go/discovery/metric.go [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+	return m.Timestamp == m2.Timestamp &&
+		m.InstanceKey == m2.InstanceKey &&
+		m.BackendLatency == m2.BackendLatency &&
+		m.InstanceLatency == m2.InstanceLatency &&
+		m.TotalLatency == m2.TotalLatency &&
+		m.Err == m2.Err
+}
+
+// MetricsEqual compares two slices of Metrics to see if they are the same
+func MetricsEqual(m1, m2 [](*Metric)) bool {
```
Comment (shlomi-noach) [inline go/discovery/metric.go]: You can use https://golang.org/pkg/reflect/#DeepEqual

**Review by shlomi-noach — CHANGES_REQUESTED** [review]
See questions within review.

I used a lot of :+1: for self-marking pieces of code I've reviewed.

I wish to be able to turn this feature off. Id expect to see a `if DiscoveryCollectionRetentionSeconds > 0 { // do it }`, but it seems to me like this is always on.

File: go/discovery/metric_json.go [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/github/orchestrator/go/collection"
+)
+
+// myfloat is to force the JSON output to show 3 decimal places
+type myfloat float64
```
Comment (shlomi-noach) [inline go/discovery/metric_json.go]: Please rename to `formattedFloat` or something

File: go/discovery/metric_json.go [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+			Port:      m.InstanceKey.Port,
+			BackendLatencySeconds:  myfloat(m.BackendLatency.Seconds()),
+			InstanceLatencySeconds: myfloat(m.InstanceLatency.Seconds()),
+			TotalLatencySeconds:    myfloat(m.TotalLatency.Seconds()),
+			Err:                    m.Err,
+		}
+		s = append(s, mj)
+	}
+	return s, nil
+}
```
Comment (shlomi-noach) [inline go/discovery/metric_json.go:74]: :+1:

File: go/http/api.go [hunk 23]
```diff
@@ -1595,6 +1600,40 @@ func (this *HttpAPI) BulkInstances(params martini.Params, r render.Render, req *
...
+	refTime := time.Now().Add(-time.Duration(seconds) * time.Second)
+	json, err := discovery.JSONSince(discoveryMetrics, refTime)
+	if err != nil {
+		r.JSON(200, &APIResponse{Code: ERROR, Message: "Unable to determine start time. Perhaps seconds value is wrong?"})
+		return
+	}
+	log.Debugf("DiscoveryMetricsRaw data: retrieved %d entries from discovery.MC", len(json))
+
+	r.JSON(200, json)
+}
```
Comment (shlomi-noach) [inline go/http/api.go:1620]: :+1:

File: go/http/api.go [hunk 24]
```diff
@@ -1595,6 +1600,40 @@ func (this *HttpAPI) BulkInstances(params martini.Params, r render.Render, req *
...
+	refTime := time.Now().Add(-time.Duration(seconds) * time.Second)
+	aggregated, err := discovery.AggregatedSince(discoveryMetrics, refTime)
+	if err != nil {
+		r.JSON(200, &APIResponse{Code: ERROR, Message: "Unable to generate aggregated discovery metrics"})
+		return
+	}
+	// log.Debugf("DiscoveryMetricsAggregated data: %+v", aggregated)
+	r.JSON(200, aggregated)
+}
+
```
Comment (shlomi-noach) [inline go/http/api.go:1636]: :+1:

File: go/inst/instance_dao.go [hunk 25]
```diff
@@ -84,8 +91,28 @@ func InitializeInstanceDao() {
...
+			if s, ok := r.(string); ok {
+				m.Err = errors.New(s)
+			} else {
+				m.Err = r.(error)
+			}
+		}
+		m.ExecuteLatency = time.Since(m.Timestamp.Add(m.WaitLatency))
+		backendWrites.Append(m)
+		<-instanceWriteChan // assume this takes no time
+	}()
```
Comment (shlomi-noach) [inline go/inst/instance_dao.go:115]: :+1:

File: go/inst/instance_dao.go [hunk 26]
```diff
@@ -541,38 +662,46 @@ func ReadTopologyInstanceBufferable(instanceKey *InstanceKey, bufferWrites bool)
...
 
 	// Something is wrong, could be network-wise. Record that we
 	// tried to check the instance. last_attempted_check is also
 	// updated on success by writeInstance.
+	latency.Start("backend")
 	_ = UpdateInstanceLastAttemptedCheck(instanceKey)
 	_ = UpdateInstanceLastChecked(&instance.Key)
-	return nil, fmt.Errorf("Failed ReadTopologyInstance")
+	latency.Stop("backend")
+	return nil, fmt.Errorf("ReadTopologyInstanceBufferable failed: %v", err)
```
Comment (shlomi-noach) [inline go/inst/instance_dao.go:704]: :+1:

File: go/inst/instance_dao.go [hunk 27]
```diff
@@ -710,10 +843,9 @@ func ReadInstancePromotionRule(instance *Instance) (err error) {
 	}
 
 	var promotionRule CandidatePromotionRule = NeutralPromoteRule
-	// Read the cluster_name of the _master_ of our instance, derive it from there.
 	query := `
 			select
-					ifnull(nullif(promotion_rule, ''), 'neutral') as promotion_rule
+				ifnull(nullif(promotion_rule, ''), 'neutral') as promotion_rule
```
Comment (shlomi-noach) [inline go/inst/instance_dao.go:848]: :+1:

File: go/logic/orchestrator.go [hunk 28]
```diff
@@ -112,12 +125,21 @@ func handleDiscoveryRequests() {
...
+// replicas (if any) are also checked.
 func discoverInstance(instanceKey inst.InstanceKey) {
-	start := time.Now()
+	// create stopwatch entries
+	latency := stopwatch.NewNamedStopwatch()
+	latency.AddMany([]string{
+		"backend",
+		"instance",
+		"total"})
+	latency.Start("total") // start the total stopwatch (not changed anywhere else)
```
Comment (shlomi-noach) [inline go/logic/orchestrator.go:138]: :+1:

**Review by shlomi-noach — CHANGES_REQUESTED** [review]

File: go/discovery/metric.go [hunk 29]
```diff
@@ -39,29 +40,13 @@ func (m Metric) When() time.Time {
...
-	return m.Timestamp == m2.Timestamp &&
-		m.InstanceKey == m2.InstanceKey &&
-		m.BackendLatency == m2.BackendLatency &&
-		m.InstanceLatency == m2.InstanceLatency &&
-		m.TotalLatency == m2.TotalLatency &&
-		m.Err == m2.Err
-}
-
 // MetricsEqual compares two slices of Metrics to see if they are the same
 func MetricsEqual(m1, m2 [](*Metric)) bool {
```
Comment (shlomi-noach) [inline go/discovery/metric.go]: `return reflect.DeepEqual(m1,m2)`

File: go/discovery/metric.go [hunk 30]
```diff
@@ -39,29 +40,13 @@ func (m Metric) When() time.Time {
...
 	if len(m1) != len(m2) {
 		return false
 	}
 	for i := range m1 {
-		if !m1[i].Equal(m2[i]) {
+		if !reflect.DeepEqual(m1[i],m2[i]) {
```
Comment (shlomi-noach) [inline go/discovery/metric.go]: You want to use `DeepEqual` on everything, not on a per-element basis.

**Review by shlomi-noach — CHANGES_REQUESTED** [review]
You misunderstood my comment about `DeepEqual`

**Comment by shlomi-noach** [comment]
https://github.com/github/orchestrator/pull/79#discussion_r101912521

File: go/logic/orchestrator.go [hunk 31]
```diff
@@ -168,41 +168,45 @@ func discoverInstance(instanceKey inst.InstanceKey) {
...
 			Err:             err,
 		})
 		log.Warningf("discoverInstance(%+v) instance is nil in %.3fs (Backend: %.3fs, Instance: %.3fs), error=%+v",
 			instanceKey,
-			latency.ElapsedSeconds("total"),
-			latency.ElapsedSeconds("backend"),
-			latency.ElapsedSeconds("instance"),
+			totalLatency,
+			backendLatency,
+			instanceLatency,
```
Comment (shlomi-noach) [inline go/logic/orchestrator.go]: :+1:

**Comment by shlomi-noach** [comment]
@sjmudd the last outstanding issue is about DeepEquals, see https://github.com/github/orchestrator/pull/79#discussion_r101912521

File: go/discovery/metric_json.go (hunk 21 above)
Comment (sjmudd) [reply go/discovery/metric_json.go]: done.

**Comment by sjmudd** [comment]
I think all your concerns are now addressed. Not sure how to grab a quick view of anything that's pending. If I've missed something please let me know. Thanks.

**Review by shlomi-noach — APPROVED** [review]

**Comment by sjmudd** [comment]
+1 Thanks for taking this.
